zkevm-circuits/src/evm_circuit/execution/error_oog_call.rs

```diff
@@ -0,0 +1,511 @@
+    callee_code_hash: Cell<F>,
+    is_empty_nonce_and_balance: BatchedIsZeroGadget<F, 2>,
+    is_empty_code_hash: IsEqualGadget<F>,
+    insufficient_gas: LtGadget<F, N_BYTES_GAS>,
+    restore_context: RestoreContextGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ErrorOOGCallGadget<F> {
+    const NAME: &'static str = "ErrorOutOfGasCall";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::ErrorOutOfGasCALL;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+        cb.opcode_lookup(opcode.expr(), 1.expr());
+        // TODO: add CallCode etc. when handle ErrorOutOfGasCALLCODE in furture
+        // implementation
+        cb.require_equal(
+            "ErrorOutOfGasCall opcode is Call",
+            opcode.expr(),
+            OpcodeId::CALL.expr(),
+        );
+
+        let gas_word = cb.query_word();
+        let callee_address_word = cb.query_word();
+        let value = cb.query_word();
+        let cd_offset = cb.query_cell();
+        let cd_length = cb.query_rlc();
+        let rd_offset = cb.query_cell();
+        let rd_length = cb.query_rlc();
+        let balance = cb.query_word();
+
+        let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
+
+        let mut reversion_info = cb.reversion_info_read(None);
+        let [is_static] =
+            [CallContextFieldTag::IsStatic].map(|field_tag| cb.call_context(None, field_tag));
+
+        // Lookup values from stack
+        cb.stack_pop(gas_word.expr());
+        cb.stack_pop(callee_address_word.expr());
+        cb.stack_pop(value.expr());
+        cb.stack_pop(cd_offset.expr());
+        cb.stack_pop(cd_length.expr());
+        cb.stack_pop(rd_offset.expr());
+        cb.stack_pop(rd_length.expr());
+        cb.stack_push(0.expr());
+
+        // Recomposition of random linear combination to integer
+        let callee_address =
+            from_bytes::expr(&callee_address_word.cells[..N_BYTES_ACCOUNT_ADDRESS]);
+        let cd_address = MemoryAddressGadget::construct(cb, cd_offset, cd_length);
+        let rd_address = MemoryAddressGadget::construct(cb, rd_offset, rd_length);
+        let memory_expansion =
+            MemoryExpansionGadget::construct(cb, [cd_address.address(), rd_address.address()]);
+
+        // Add callee to access list
+        let is_warm = cb.query_bool();
+        let is_warm_prev = cb.query_bool();
+        cb.account_access_list_write(
+            tx_id.expr(),
+            callee_address.clone(),
+            is_warm.expr(),
+            is_warm_prev.expr(),
+            Some(&mut reversion_info),
+        );
+
+        let value_is_zero = IsZeroGadget::construct(cb, sum::expr(&value.cells));
+        let has_value = 1.expr() - value_is_zero.expr();
+        cb.condition(has_value.clone(), |cb| {
+            cb.require_zero(
+                "CALL with value must not be in static call stack",
+                is_static.expr(),
+            );
+        });
```

Comment:
for `ErrorOutOfGasSTATICCALL` I am planing to merge it into `oog_call` in the future, at the moment we can remove it.

Reply:
I mean even we consider out of gas for `STATICCALL`, it's still not necessary.

Out of gas and violation of static context are 2 different kinds of exceptions, and in go-ethereum implementation out of gas will always be detected before the later one, so this will become an over constraint if they happen at the same time.

Reply:
Got it , that is true.

Reply:
remove it in [7ca4a2f](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/827/commits/7ca4a2fa8ddc5f740be472d36309b5323495f0e9) thanks !